bpf/gotracer/go_sql.c

```diff
@@ -25,11 +25,119 @@
 #include <common/ringbuf.h>
 
 #include <gotracer/go_common.h>
+#include <gotracer/go_str.h>
 
-static __always_inline void set_sql_info(void *goroutine_addr, void *sql_param, void *query_len) {
+// Validates that driverConn.ci points to a MySQL connection and returns the mysqlConn pointer.
+static __always_inline void *get_mysql_conn_ptr(u64 driver_conn_ptr) {
+    if (driver_conn_ptr == 0) {
+        return NULL;
+    }
+
+    off_table_t *ot = get_offsets_table();
+
+    // Get driverConn.ci offset
+    u64 ci_offset = go_offset_of(ot, (go_offset){.v = _driverconn_ci_pos});
+    if (!ci_offset) {
+        bpf_dbg_printk("can't get driverConn.ci offset");
+        return NULL;
+    }
+
+    // driverConn.ci is a Go interface [type_ptr (8 bytes), data_ptr (8 bytes)]
+    // Read the type pointer (at ci_offset + 0) to validate driver type
+    void *ci_type_ptr = 0;
+    int res =
+        bpf_probe_read(&ci_type_ptr, sizeof(ci_type_ptr), (void *)(driver_conn_ptr + ci_offset));
+
+    if (res || !ci_type_ptr) {
+        bpf_dbg_printk("can't read driverConn.ci type pointer");
+        return NULL;
+    }
+
+    u64 mysql_type_addr = go_offset_of(ot, (go_offset){.v = _mysql_conn_type_off});
+    if (!mysql_type_addr) {
+        bpf_dbg_printk("can't read mysql.mysqlConn offset");
+        return NULL;
+    }
+
+    bpf_dbg_printk("validating mysql conn type %llx with %llx", mysql_type_addr, ci_type_ptr);
+    if ((u64)ci_type_ptr != mysql_type_addr) {
+        bpf_dbg_printk("connection type doesn't match from mysql.mysqlConn");
+        return NULL;
+    }
+
+    void *mysql_conn_ptr = 0;
+    res = bpf_probe_read(
+        &mysql_conn_ptr, sizeof(mysql_conn_ptr), (void *)(driver_conn_ptr + ci_offset + 8));
+
+    if (res || !mysql_conn_ptr) {
+        bpf_dbg_printk("can't read MySQL connection data pointer");
+        return NULL;
+    }
+
+    return mysql_conn_ptr;
+}
+
+// Extracts MySQL server hostname from a validated mysqlConn pointer.
+// Follows the pointer chain: mysqlConn -> cfg (*Config) -> Addr (string)
+static __always_inline int
+read_mysql_hostname_from_mysqlconn(void *mysql_conn_ptr, char *hostname, int max_len) {
```

Comment:
```suggestion
read_mysql_hostname_from_mysqlconn(const void *mysql_conn_ptr, char *hostname, int max_len) {
```

rationale: this pointer is immutable in this function (if you check the prototype of bpf_probe_read_* you will also see that it takes a `const void*` based on the same rationale).